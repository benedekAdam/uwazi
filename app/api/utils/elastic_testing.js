import entities from 'api/entities/entities';
import elasticMapping from '../../../database/elastic_mapping';
import elasticConfig from 'api/config/elasticIndexes';
import elastic from 'api/search/elastic';

export default (elasticIndex) => {
  elasticConfig.index = elasticIndex;
  return {
    resetIndex() {
      return elastic.indices.delete({ index: elasticIndex, ignore_unavailable: true })
      .then(() => elastic.indices.create({
          index: elasticIndex,
          body: elasticMapping
      }))
      .then(() => null);
    },
    reindex() {
      return this.resetIndex()
      .then(() => entities.indexEntities({}, '+fullText'))
      .then(() => elastic.indices.refresh({ index: elasticIndex }))
      .then(() => null);
    },

    refresh() {
      return elastic.indices.refresh({ index: elasticIndex });
    }
  };
};
