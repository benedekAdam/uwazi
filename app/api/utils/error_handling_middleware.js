import errorLog from 'api/log/errorLog';

export default function (req, res, next) {
  res.error = (error) => {
    let result = error;
    if (error instanceof Error) {
      result = error.stack.split('\n');
    }

    if (error.code) {
      result = error.message;
    }

    let code = error.code || 500;

    if (!(error.code > 0) || error.name === 'MongoError') {
      code = 500;
    }

    res.status(code);
    res.json({ error: result });

    if (error.code === 500) {
      errorLog.error(result);
    }
  };
  next();
}
