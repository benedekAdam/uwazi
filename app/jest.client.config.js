module.exports = {
  name: 'client',
  displayName: 'Client',
  testMatch: [
    '**/react/**/specs/*spec.js?(x)'
  ],
  testPathIgnorePatterns: [],
  setupTestFrameworkScriptFile: '<rootDir>/setUpJestClient.js',
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy'
  },
  snapshotSerializers: ['enzyme-to-json/serializer']
};
