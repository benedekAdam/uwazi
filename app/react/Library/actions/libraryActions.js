import * as types from 'app/Library/actions/actionTypes';
import api from 'app/Search/SearchAPI';
import { notify } from 'app/Notifications';
import { actions as formActions } from 'react-redux-form';
import { actions } from 'app/BasicReducer';
import documents from 'app/Documents';
import { browserHistory } from 'react-router';
import rison from 'rison';
import referencesAPI from 'app/Viewer/referencesAPI';
import { api as entitiesAPI } from 'app/Entities';
import { toUrlParams } from 'shared/JSONRequest';

export function enterLibrary() {
  return { type: types.ENTER_LIBRARY };
}

export function initializeFiltersForm(values = {}) {
  return Object.assign(values, { type: types.INITIALIZE_FILTERS_FORM });
}

export function selectDocument(doc) {
  let document = doc;
  if (doc.toJS) {
    document = doc.toJS();
  }
  return { type: types.SELECT_DOCUMENT, doc: document };
}

export function getAndSelectDocument(id) {
  return (dispatch) => {
    entitiesAPI.get(id)
    .then((entity) => {
      dispatch({ type: types.SELECT_SINGLE_DOCUMENT, doc: entity[0] });
    });
  };
}

export function selectDocuments(docs) {
  return { type: types.SELECT_DOCUMENTS, docs };
}

export function unselectDocument(docId) {
  return { type: types.UNSELECT_DOCUMENT, docId };
}

export function selectSingleDocument(doc) {
  return { type: types.SELECT_SINGLE_DOCUMENT, doc };
}

export function unselectAllDocuments() {
  return { type: types.UNSELECT_ALL_DOCUMENTS };
}

export function updateSelectedEntities(entities) {
  return { type: types.UPDATE_SELECTED_ENTITIES, entities };
}

export function showFilters() {
  return { type: types.SHOW_FILTERS };
}

export function hideFilters() {
  return { type: types.HIDE_FILTERS };
}

export function setDocuments(docs) {
  return { type: types.SET_DOCUMENTS, documents: docs };
}

export function unsetDocuments() {
  return { type: types.UNSET_DOCUMENTS };
}

export function setTemplates(templates, thesauris) {
  return (dispatch) => {
    dispatch({ type: types.SET_LIBRARY_TEMPLATES, templates, thesauris });
  };
}

export function setPreviewDoc(docId) {
  return { type: types.SET_PREVIEW_DOC, docId };
}

export function setSuggestions(suggestions) {
  return { type: types.SET_SUGGESTIONS, suggestions };
}

export function hideSuggestions() {
  return { type: types.HIDE_SUGGESTIONS };
}

export function showSuggestions() {
  return { type: types.SHOW_SUGGESTIONS };
}

export function setOverSuggestions(boolean) {
  return { type: types.OVER_SUGGESTIONS, hover: boolean };
}

export function filterIsEmpty(value) {
  if (value && value.values && !value.values.length) {
    return true;
  }

  if (Array.isArray(value) && !value.length) {
    return true;
  }

  if (typeof value === 'string' && !value) {
    return true;
  }

  if (typeof value === 'object') {
    const hasValue = Object.keys(value).reduce((result, key) => result || Boolean(value[key]), false);
    return !hasValue;
  }

  return false;
}

export function processFilters(readOnlySearch, filters, limit) {
  const search = Object.assign({ filters: {} }, readOnlySearch);
  search.filters = {};

  filters.properties.forEach((property) => {
    if (!filterIsEmpty(readOnlySearch.filters[property.name]) && !property.filters) {
      search.filters[property.name] = readOnlySearch.filters[property.name];
    }

    if (property.filters) {
      const searchFilter = Object.assign({}, readOnlySearch.filters[property.name]);
      property.filters.forEach((filter) => {
        if (filterIsEmpty(searchFilter[filter.name])) {
          delete searchFilter[filter.name];
        }
      });

      if (Object.keys(searchFilter).length) {
        search.filters[property.name] = searchFilter;
      }
    }
  });

  search.types = filters.documentTypes;
  search.limit = limit;
  return search;
}

export function encodeSearch(search, appendQ = true) {
  Object.keys(search).forEach((key) => {
    if (search[key] && search[key].length === 0) {
      delete search[key];
    }

    if (typeof search[key] === 'object' && Object.keys(search[key]).length === 0) {
      delete search[key];
    }

    if (search[key] === '') {
      delete search[key];
    }
  });

  return appendQ ? `?q=${rison.encode(search)}` : rison.encode(search);
}

export function searchDocuments({ search, filters }, storeKey, limit) {
  return (dispatch, getState) => {
    const state = getState()[storeKey];
    let currentFilters;
    if (filters) {
      currentFilters = filters;
    }
    if (filters && filters.toJS) {
      currentFilters = filters.toJS();
    }
    if (!currentFilters) {
      currentFilters = state.filters.toJS();
    }

    const finalSearchParams = processFilters(search, currentFilters, limit);
    finalSearchParams.searchTerm = state.search.searchTerm;


    const currentSearch = browserHistory.getCurrentLocation().q || '()';
    const currentSearchParams = rison.decode(decodeURIComponent(currentSearch));
    if (finalSearchParams.searchTerm && finalSearchParams.searchTerm !== currentSearchParams.searchTerm) {
      finalSearchParams.sort = '_score';
    }

    if (search.userSelectedSorting) {
      dispatch(actions.set(`${storeKey}.selectedSorting`, search));
    }

    const { pathname } = browserHistory.getCurrentLocation();
    const path = (`${pathname}/`).replace(/\/\//g, '/');
    const query = browserHistory.getCurrentLocation().query || {};

    query.q = encodeSearch(finalSearchParams, false);
    browserHistory.push(path + toUrlParams(query));
  };
}

export function elementCreated(doc) {
  return { type: types.ELEMENT_CREATED, doc };
}

export function updateEntity(updatedDoc) {
  return { type: types.UPDATE_DOCUMENT, doc: updatedDoc };
}

export function updateEntities(updatedDocs) {
  return { type: types.UPDATE_DOCUMENTS, docs: updatedDocs };
}

export function searchSnippets(searchTerm, sharedId, storeKey) {
  return dispatch => api.searchSnippets(searchTerm, sharedId)
  .then((snippets) => {
    dispatch(actions.set(`${storeKey}.sidepanel.snippets`, snippets));
    return snippets;
  });
}

export function saveDocument(doc, formKey) {
  return dispatch => documents.api.save(doc)
  .then((updatedDoc) => {
    dispatch(notify('Document updated', 'success'));
    dispatch(formActions.reset(formKey));
    dispatch(updateEntity(updatedDoc));
    dispatch(selectSingleDocument(updatedDoc));
  });
}

export function multipleUpdate(entities, values) {
  return (dispatch) => {
    const updatedEntities = entities.toJS().map((entity) => {
      entity.metadata = Object.assign({}, entity.metadata, values.metadata);
      if (values.icon) {
        entity.icon = values.icon;
      }
      return entity;
    });

    const updatedEntitiesIds = updatedEntities.map(entity => entity.sharedId);
    return entitiesAPI.multipleUpdate(updatedEntitiesIds, values)
    .then(() => {
      dispatch(notify('Update success', 'success'));
      dispatch(updateEntities(updatedEntities));
    });
  };
}

export function saveEntity(entity, formModel) {
  return dispatch => entitiesAPI.save(entity)
  .then((updatedDoc) => {
    dispatch(formActions.reset(formModel));
    dispatch(unselectAllDocuments());
    if (entity._id) {
      dispatch(notify('Entity updated', 'success'));
      dispatch(updateEntity(updatedDoc));
    } else {
      dispatch(notify('Entity created', 'success'));
      dispatch(elementCreated(updatedDoc));
    }

    dispatch(selectSingleDocument(updatedDoc));
  });
}

export function removeDocument(doc) {
  return { type: types.REMOVE_DOCUMENT, doc };
}

export function removeDocuments(docs) {
  return { type: types.REMOVE_DOCUMENTS, docs };
}

export function deleteDocument(doc) {
  return dispatch => documents.api.delete(doc)
  .then(() => {
    dispatch(notify('Document deleted', 'success'));
    dispatch(unselectAllDocuments());
    dispatch(removeDocument(doc));
  });
}

export function deleteEntity(entity) {
  return dispatch => entitiesAPI.delete(entity)
  .then(() => {
    dispatch(notify('Entity deleted', 'success'));
    dispatch(unselectDocument(entity._id));
    dispatch(removeDocument(entity));
  });
}

export function loadMoreDocuments(storeKey, amount) {
  return (dispatch, getState) => {
    searchDocuments({ search: getState()[storeKey].search }, storeKey, amount)(dispatch, getState);
  };
}

export function getSuggestions() {
  return { type: 'GET_SUGGESTIONS' };
}

export function getDocumentReferences(documentId, storeKey) {
  return dispatch => referencesAPI.get(documentId)
  .then((references) => {
    dispatch(actions.set(`${storeKey}.sidepanel.references`, references));
  });
}
