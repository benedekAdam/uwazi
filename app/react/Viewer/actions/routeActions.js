import { actions } from 'app/BasicReducer';
import { getDocument } from 'app/Viewer/actions/documentActions';
import referencesAPI from 'app/Viewer/referencesAPI';
import relationTypesAPI from 'app/RelationTypes/RelationTypesAPI';
import * as relationships from 'app/Relationships/utils/routeUtils';

import { setReferences } from './referencesActions';
import entitiesAPI from '../../Entities/EntitiesAPI';

export function requestViewerState({ documentId, raw, page }, globalResources) {
  return Promise.all([
    getDocument(documentId),
    referencesAPI.get(documentId),
    relationTypesAPI.get(),
    relationships.requestState(documentId, globalResources.templates),
    raw ? entitiesAPI.getRawPage(documentId, page) : ''
  ])
  .then(([doc, references, relationTypes, [connectionsGroups, searchResults, sort], rawText]) => ({
    documentViewer: {
      doc,
      references,
      relationTypes,
      rawText,
    },
    relationships: {
      list: {
        entityId: doc.sharedId,
        entity: doc,
        connectionsGroups,
        searchResults,
        sort,
        filters: {},
        view: 'graph'
      }
    },
    relationTypes,
  }));
}

export function setViewerState(state) {
  return (dispatch) => {
    const { documentViewer } = state;
    dispatch(actions.set('relationTypes', state.relationTypes));
    dispatch(actions.set('viewer/doc', documentViewer.doc));
    dispatch(actions.set('viewer/relationTypes', documentViewer.relationTypes));
    dispatch(actions.set('viewer/rawText', documentViewer.rawText));
    dispatch(setReferences(documentViewer.references));
    dispatch(relationships.setReduxState(state));
  };
}
