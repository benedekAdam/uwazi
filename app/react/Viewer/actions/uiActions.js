import { actions } from 'app/BasicReducer';
import { events } from 'app/utils';
import { setTargetSelection } from 'app/Viewer/actions/selectionActions';
import Marker from 'app/Viewer/utils/Marker.js';
import scroller from 'app/Viewer/utils/Scroller';
import * as types from 'app/Viewer/actions/actionTypes';

export function closePanel() {
  return {
    type: types.CLOSE_PANEL
  };
}

export function openPanel(panel) {
  return {
    type: types.OPEN_PANEL,
    panel
  };
}

export function resetReferenceCreation() {
  return function (dispatch) {
    dispatch({ type: types.RESET_REFERENCE_CREATION });
    dispatch(actions.unset('viewer/targetDoc'));
    dispatch(actions.unset('viewer/targetDocHTML'));
    dispatch(actions.unset('viewer/targetDocReferences'));
  };
}

export function selectTargetDocument(id) {
  return {
    type: types.SELECT_TARGET_DOCUMENT,
    id
  };
}

export function highlightReference(reference) {
  return {
    type: types.HIGHLIGHT_REFERENCE,
    reference
  };
}

export function deactivateReference() {
  return {
    type: types.DEACTIVATE_REFERENCE
  };
}

export function goToActive(value = true) {
  return {
    type: types.GO_TO_ACTIVE,
    value
  };
}

export function highlightSnippets(snippets, pagesBeingRendered = []) {
  const highlights = snippets.get('fullText')
  .filter(s => pagesBeingRendered.includes(s.get('page')))
  .map(snippet => snippet.get('text')
  .replace(/[-[\]{}()*+?.,\\^$|#]/g, '\\$&')
  .replace(/<[^>]*>/g, '')
  .replace(/\s+/g, '\\s*')
  .replace(/\n/g, '\\s*'))
  .filter((elem, pos, arr) => arr.indexOf(elem) === pos);

  Marker.unmark();
  highlights.forEach((highlightRegExp) => {
    const regexp = new RegExp(highlightRegExp);
    Marker.markRegExp(regexp, { separateWordSearch: false, acrossElements: true });
  });
}

export function scrollToPage(page, duration = 100) {
  scroller.to(`.document-viewer div#page-${page}`, '.document-viewer', { duration, dividerOffset: 1 });
}

export function scrollTo(reference, docInfo, element = 'a') {
  //
  const page = Object.keys(docInfo).find(pageNumber => docInfo[pageNumber].chars >= reference.range.start);
  //

  if (window.document.querySelector(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer')) {
    scroller.to(`.document-viewer a[data-id="${reference._id}"]`, '.document-viewer', { duration: 100 });
  } else {
    const scroll = scroller.to(`.document-viewer div#page-${page}`, '.document-viewer', { duration: 0, dividerOffset: 1 });

    events.on('referenceRendered', (renderedReference) => {
      if (renderedReference._id === reference._id &&
          window.document.querySelector(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer')
      ) {
        window.clearInterval(scroll);
        scroller.to(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer', { duration: 100 });
        events.removeAllListeners('referenceRendered');
      }
    });
  }

  scroller.to(`.metadata-sidepanel .item-${reference._id}`, '.metadata-sidepanel .sidepanel-body', { duration: 100 });
}

export function activateReference(reference, docInfo, tab) {
  const tabName = tab && !Array.isArray(tab) ? tab : 'references';
  events.removeAllListeners('referenceRendered');

  return function (dispatch) {
    dispatch({ type: types.ACTIVE_REFERENCE, reference: reference._id });
    dispatch({ type: types.OPEN_PANEL, panel: 'viewMetadataPanel' });
    dispatch(actions.set('viewer.sidepanel.tab', tabName));

    setTimeout(() => {
      scrollTo(reference, docInfo);
    });
  };
}

export function scrollToActive(reference, docInfo, tab, doScroll) {
  return function (dispatch) {
    if (doScroll) {
      dispatch(goToActive(false));
      dispatch(activateReference(reference, docInfo, tab));
    }
  };
}


export function selectReference(reference, docInfo) {
  return function (dispatch) {
    dispatch(activateReference(reference, docInfo));
    dispatch(setTargetSelection(reference.range));
  };
}
