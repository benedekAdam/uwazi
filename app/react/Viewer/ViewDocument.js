import { Helmet } from 'react-helmet';
import { browserHistory } from 'react-router';
import { actions as formActions } from 'react-redux-form';
import React from 'react';

import { actions } from 'app/BasicReducer';
import { isClient } from 'app/utils';
import { setReferences } from 'app/Viewer/actions/referencesActions';
import { toUrlParams } from 'shared/JSONRequest';
import RouteHandler from 'app/App/RouteHandler';
import Viewer from 'app/Viewer/components/Viewer';
import entitiesAPI from 'app/Entities/EntitiesAPI';
import * as relationships from 'app/Relationships/utils/routeUtils';
import { scrollToPage } from './actions/uiActions';

import { requestViewerState, setViewerState } from './actions/routeActions';

class ViewDocument extends RouteHandler {
  constructor(props, context) {
    //Force client state even if is rendered from server to force the pdf character count process
    RouteHandler.renderedFromServer = props.renderedFromServer || false;
    //
    super(props, context);
    this.changeBrowserHistoryPage = this.changeBrowserHistoryPage.bind(this);
    this.changePage = this.changePage.bind(this);
    this.onDocumentReady = this.onDocumentReady.bind(this);
  }

  static requestState(routeParams, query = {}, globalResources) {
    return requestViewerState({ ...routeParams, raw: (query.raw === 'true') || !isClient, page: query.page }, globalResources);
  }

  componentWillUnmount() {
    this.emptyState();
  }

  componentWillMount() {
    if (this.props.location.query.searchTerm) {
      this.context.store.dispatch(actions.set('viewer.sidepanel.tab', 'text-search'));
    }
  }

  componentWillReceiveProps(props) {
    super.componentWillReceiveProps(props);
    const { query = {} } = props.location;
    if ((query.page !== this.props.location.query.page || query.raw !== this.props.location.query.raw) && query.raw === 'true') {
      return entitiesAPI.getRawPage(props.params.documentId, query.page)
      .then((pageText) => {
        this.context.store.dispatch(actions.set('viewer/rawText', pageText));
      });
    }
  }

  emptyState() {
    this.context.store.dispatch(actions.unset('viewer/doc'));
    this.context.store.dispatch(actions.unset('viewer/templates'));
    this.context.store.dispatch(actions.unset('viewer/thesauris'));
    this.context.store.dispatch(actions.unset('viewer/relationTypes'));
    this.context.store.dispatch(actions.unset('viewer/rawText'));
    this.context.store.dispatch(formActions.reset('documentViewer.tocForm'));
    this.context.store.dispatch(actions.unset('viewer/targetDoc'));
    this.context.store.dispatch(setReferences([]));
    this.context.store.dispatch(relationships.emptyState());
  }

  setReduxState(state) {
    this.context.store.dispatch(setViewerState(state));
  }

  changePage(nextPage) {
    if (!this.props.location.query.raw) {
      return scrollToPage(nextPage);
    }

    return this.changeBrowserHistoryPage(nextPage);
  }

  changeBrowserHistoryPage(newPage) {
    const { query: { page, ...queryWithoutPage } } = this.props.location;
    queryWithoutPage.raw = queryWithoutPage.raw || undefined;
    browserHistory.push(`${this.props.location.pathname}${toUrlParams({ ...queryWithoutPage, page: newPage })}`);
  }

  onDocumentReady() {
    if (this.props.location.query.raw !== 'true' && this.props.location.query.page) {
      scrollToPage(this.props.location.query.page, 0);
    }
  }

  render() {
    const { query = {}, pathname } = this.props.location;
    const raw = query.raw === 'true' || !isClient;
    const page = Number(query.page || 1);
    return (
      <React.Fragment>
        <Helmet>
          {raw && <link rel="canonical" href={`${pathname}?page=${page}`} />}
        </Helmet>
        <Viewer
          raw={raw}
          searchTerm={query.searchTerm}
          onPageChange={this.changeBrowserHistoryPage}
          onDocumentReady={this.onDocumentReady}
          changePage={this.changePage}
        />
      </React.Fragment>
    );
  }
}

ViewDocument.defaultProps = {
  params: {}
};

export default ViewDocument;
