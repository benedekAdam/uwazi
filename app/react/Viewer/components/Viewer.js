import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { ConnectionsList } from 'app/ConnectionsList';
import { CreateConnectionPanel } from 'app/Connections';
import { CurrentLocationLink } from 'app/Layout';
import { RelationshipsFormButtons } from 'app/Relationships';
import { Translate } from 'app/I18N';
import { actions } from 'app/BasicReducer';
import AddEntitiesPanel from 'app/Relationships/components/AddEntities';
import ContextMenu from 'app/ContextMenu';
import Footer from 'app/App/Footer';
import Marker from 'app/Viewer/utils/Marker';
import RelationshipMetadata from 'app/Relationships/components/RelationshipMetadata';
import ShowIf from 'app/App/ShowIf';

import { PaginatorWithPage } from './Paginator';
import { addReference } from '../actions/referencesActions';
import { loadDefaultViewerMenu, loadTargetDocument } from '../actions/documentActions';
import { openPanel } from '../actions/uiActions';
import { selectDoc } from '../selectors';
import ConfirmCloseForm from './ConfirmCloseForm';
import SourceDocument from './SourceDocument';
import TargetDocument from './TargetDocument';
import ViewMetadataPanel from './ViewMetadataPanel';
import ViewerDefaultMenu from './ViewerDefaultMenu';
import ViewerTextSelectedMenu from './ViewerTextSelectedMenu';

export class Viewer extends Component {
  constructor(props) {
    super(props);
    this.state = { firstRender: true };
    this.handlePlainTextClick = this.handlePlainTextClick.bind(this);
  }

  componentWillMount() {
    this.context.store.dispatch(openPanel('viewMetadataPanel'));
    if (this.props.sidepanelTab === 'connections') {
      this.context.store.dispatch(actions.set('viewer.sidepanel.tab', ''));
    }
  }

  handlePlainTextClick() {
    this.props.showTab('metadata');
  }

  componentDidMount() {
    this.context.store.dispatch(loadDefaultViewerMenu());
    Marker.init('div.main-wrapper');
    this.setState({ firstRender: false }); // eslint-disable-line react/no-did-mount-set-state
  }


  render() {
    const { doc, sidepanelTab } = this.props;

    let className = 'document-viewer';
    if (this.props.panelIsOpen) {
      className += ' with-panel is-active';
    }
    if (this.props.targetDoc) {
      className += ' show-target-document';
    }
    if (this.props.showConnections) {
      className += ' connections';
    }

    const { raw, searchTerm, pageText } = this.props;
    return (
      <div className="row">
        <Helmet title={doc.get('title') ? doc.get('title') : 'Document'} />
        <ShowIf if={!this.props.targetDoc}>
          <div className="content-header content-header-document">
            <div className="content-header-title">
              {sidepanelTab !== 'connections'
                  &&
                  <React.Fragment>
                    <PaginatorWithPage
                      totalPages={doc.get('totalPages')}
                      onPageChange={this.props.changePage}
                    />
                    <CurrentLocationLink
                      onClick={!raw ? this.handlePlainTextClick : () => {}}
                      className="btn btn-default"
                      queryParams={{ raw: raw || this.state.firstRender ? '' : 'true' }}
                    >
                      { raw || this.state.firstRender ? <Translate>Normal view</Translate> : <Translate>Plain text</Translate> }
                    </CurrentLocationLink>
                  </React.Fragment>
              }
            </div>
          </div>
        </ShowIf>
        <main className={className}>
          <div className="main-wrapper">
            <ShowIf if={sidepanelTab !== 'connections' && !this.props.targetDoc}>
              {raw || this.state.firstRender ?
                <pre>{pageText}</pre> :
                <SourceDocument onPageChange={this.props.onPageChange} onDocumentReady={this.props.onDocumentReady}/>
              }
            </ShowIf>
            <ShowIf if={sidepanelTab === 'connections'}>
              <ConnectionsList hideFooter searchCentered />
            </ShowIf>
            <TargetDocument />
            <Footer/>
          </div>
        </main>

        <ConfirmCloseForm />
        <ViewMetadataPanel raw={raw || this.state.firstRender} storeKey="documentViewer" searchTerm={searchTerm}/>
        <CreateConnectionPanel
          containerId={this.props.targetDoc ? 'target' : doc.get('sharedId')}
          onCreate={this.props.addReference}
          onRangedConnect={this.props.loadTargetDocument}
        />

        <ShowIf if={sidepanelTab === 'connections'}>
          <RelationshipMetadata />
        </ShowIf>

        <ShowIf if={sidepanelTab === 'connections'}>
          <AddEntitiesPanel />
        </ShowIf>

        <ShowIf if={sidepanelTab === 'connections'}>
          <div className="sidepanel-footer">
            <RelationshipsFormButtons />
          </div>
        </ShowIf>

        <ContextMenu align="bottom" overrideShow show={!this.props.panelIsOpen}>
          <ViewerDefaultMenu/>
        </ContextMenu>
        <ContextMenu align="center" overrideShow show={this.props.showTextSelectMenu}>
          <ViewerTextSelectedMenu/>
        </ContextMenu>
      </div>
    );
  }
}

Viewer.defaultProps = {
  searchTerm: '',
  raw: false,
  onPageChange: () => {},
  changePage: () => {},
  onDocumentReady: () => {},
};

Viewer.propTypes = {
  searchTerm: PropTypes.string,
  raw: PropTypes.bool,
  onPageChange: PropTypes.func,
  changePage: PropTypes.func,
  onDocumentReady: PropTypes.func,
  doc: PropTypes.object,
  pageText: PropTypes.string,
  panelIsOpen: PropTypes.bool,
  addReference: PropTypes.func,
  targetDoc: PropTypes.bool,
  // TEST!!!
  sidepanelTab: PropTypes.string,
  loadTargetDocument: PropTypes.func,
  showConnections: PropTypes.bool,
  showTextSelectMenu: PropTypes.bool,
  selectedConnection: PropTypes.bool,
  selectedConnectionMetadata: PropTypes.object,
  showTab: PropTypes.func
};

Viewer.contextTypes = {
  store: PropTypes.object
};


const mapStateToProps = (state) => {
  const { documentViewer } = state;
  const uiState = documentViewer.uiState.toJS();
  return {
    pageText: documentViewer.rawText,
    doc: selectDoc(state),
    panelIsOpen: !!uiState.panel,
    targetDoc: !!documentViewer.targetDoc.get('_id'),
    // TEST!!!
    sidepanelTab: documentViewer.sidepanel.tab,
    showConnections: documentViewer.sidepanel.tab === 'references',
    showTextSelectMenu: Boolean(!documentViewer.targetDoc.get('_id') && uiState.reference && uiState.reference.sourceRange)
  };
};

const mapDispatchToProps = dispatch => bindActionCreators({
  addReference,
  loadTargetDocument,
  showTab: tab => actions.set('viewer.sidepanel.tab', tab),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Viewer);
