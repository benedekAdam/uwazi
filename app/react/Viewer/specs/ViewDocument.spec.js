import React from 'react';

import { shallow } from 'enzyme';
import ViewDocument from 'app/Viewer/ViewDocument';
import Viewer from 'app/Viewer/components/Viewer';
import * as relationships from 'app/Relationships/utils/routeUtils';

import * as routeActions from '../actions/routeActions';
import * as utils from 'app/utils';

describe('ViewDocument', () => {
  let component;
  let instance;
  let context;

  beforeEach(() => {
    const dispatch = jasmine.createSpy('dispatch');
    context = { store: { dispatch: dispatch.and.callFake((action) => {
      if (typeof action === 'function') {
        return action(dispatch);
      }
      return action;
    }) } };
    const props = {
      location: { query: {} }
    };
    component = shallow(<ViewDocument {...props} renderedFromServer />, { context });
    instance = component.instance();

    spyOn(routeActions, 'requestViewerState');

    spyOn(routeActions, 'setViewerState').and.returnValue({ type: 'setViewerState' });
  });

  it('should render the Viewer', () => {
    expect(component.find(Viewer).length).toBe(1);
  });

  describe('static requestState', () => {
    it('should call on requestViewerState', () => {
      ViewDocument.requestState({ documentId: 'documentId', lang: 'es', raw: true }, null, 'globalResources');

      expect(routeActions.requestViewerState).toHaveBeenCalledWith({ documentId: 'documentId', lang: 'es', raw: true }, 'globalResources');
    });

    it('should modify raw to true if is server side rendered', () => {
      utils.isClient = false;
      ViewDocument.requestState({ documentId: 'documentId', lang: 'es', raw: false }, null, 'globalResources');
      expect(routeActions.requestViewerState).toHaveBeenCalledWith({ documentId: 'documentId', lang: 'es', raw: true }, 'globalResources');
    });
  });


  describe('setReduxState()', () => {
    it('should dispatch setViewerState', () => {
      instance.setReduxState('state');
      expect(routeActions.setViewerState).toHaveBeenCalledWith('state');
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'setViewerState' });
    });
  });

  describe('componentWillUnmount()', () => {
    it('should call emptyState', () => {
      spyOn(instance, 'emptyState');
      instance.componentWillUnmount();

      expect(instance.emptyState).toHaveBeenCalled();
    });
  });

  describe('emptyState()', () => {
    beforeEach(() => {
      spyOn(relationships, 'emptyState').and.returnValue({ type: 'relationshipsEmptyState' });
    });

    it('should unset the state', () => {
      instance.emptyState();
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'SET_REFERENCES', references: [] });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/doc/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/templates/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/thesauris/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/relationTypes/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/rawText/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'rrf/reset', model: 'documentViewer.tocForm' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'viewer/targetDoc/UNSET' });
      expect(context.store.dispatch).toHaveBeenCalledWith({ type: 'relationshipsEmptyState' });
    });
  });
});
