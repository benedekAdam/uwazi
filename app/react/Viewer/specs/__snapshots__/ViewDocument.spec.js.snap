// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDocument when raw should render link canonical to the not raw version 1`] = `
<link
  href="pathname?page=1"
  rel="canonical"
/>
`;
